Add feature of getting condition pod log.

16yuki0702 opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes
Related issue is https://github.com/tektoncd/cli/issues/782
when running ```tkn pr logs``` without ```-f``` flag and pr tasks has conditions,
collecting condition pod logs before collecting tr pod logs.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Run the code checkers with `make check`
- [ ] Regenerate the manpages, docs and go formatting with `make generated`
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

<!--
Does your PR contain User facing changes?

If so, briefly describe them here so we can include this description in the
release notes for the next release!
-->

```
release-note
```

16yuki0702 commented:
> I have reviewed and tested this, and everything works as expected. I have just a couple comments:

> I think I would prefer the -f also be added in this pr as well so that it all goes in at once.

> Could you add a testing scenario for a failed condition in logging?
Thanks !!
I agree and address these :thumbsup:

16yuki0702 commented:
@danielhelfand sorry for late, I addresed!

I found some difference between with ```-f``` and withoud flag while making PR.

1. with ```-f```, checking pod is scheduled by ```HasScheduled``` mehod, but without flag doesn't.
2. without ```-f```, checking pod status by ```hasTaskRunFailed``` mehod, but with flag doesn't.

by these some output's difference is occured, so I wanna consider how to address on another PR.

danielhelfand commented:
I personally would still prefer that this experience be consistent across log options before the feature is introduced so there is no confusion from users. Is there perhaps a way to address the differences to make this work with `-f`?

16yuki0702 commented:
> I personally would still prefer that this experience be consistent across log options before the feature is introduced so there is no confusion from users. Is there perhaps a way to address the differences to make this work with -f?

Thanks for comments!

summarize this, the each logs showed like below

```with -f flag```
```
$ tkn p logs -f
[first-create-file : write-new-stuff] + touch /workspace/output/workspace/README.md


[file-exists : git-source-pipeline-git-hxwkm] {"level":"info","ts":1585623744.5139658,"caller":"git/git.go:102","msg":"Successfully cloned https://github.com/tektoncd/pipeline @ master in path /workspace/workspace"}
[file-exists : git-source-pipeline-git-hxwkm] {"level":"warn","ts":1585623744.5140567,"caller":"git/git.go:149","msg":"Unexpected error: creating symlink: symlink /tekton/home/.ssh /root/.ssh: file exists"}
[file-exists : git-source-pipeline-git-hxwkm] {"level":"info","ts":1585623744.5506163,"caller":"git/git.go:130","msg":"Successfully initialized and updated submodules in path /workspace/workspace"}



[file-exists : condition-check-file-exists] + test -f /workspace/workspace/README2.md


failed to get logs for task file-exists : container step-condition-check-file-exists has failed  : [{"key":"StartedAt","value":"2020-03-31T03:02:25Z","resourceRef":{}}]
```


```without -f flag```
```
$ tkn p logs

[first-create-file : write-new-stuff] + touch /workspace/output/workspace/README.md


task file-exists has failed: "step-condition-check-file-exists" exited with code 1 (image: "docker-pullable://alpine@sha256:b276d875eeed9c7d3f1cfa7edb06b22ed22b14219a7d67c52c56612330348239"); for logs run: kubectl -n default logs condtional-pr-then-check-bqdx4-file-exists-d8zzr-pod-qwqz6 -c step-condition-check-file-exists
[file-exists : git-source-pipeline-git-hxwkm] {"level":"info","ts":1585623744.5139658,"caller":"git/git.go:102","msg":"Successfully cloned https://github.com/tektoncd/pipeline @ master in path /workspace/workspace"}
[file-exists : git-source-pipeline-git-hxwkm] {"level":"warn","ts":1585623744.5140567,"caller":"git/git.go:149","msg":"Unexpected error: creating symlink: symlink /tekton/home/.ssh /root/.ssh: file exists"}
[file-exists : git-source-pipeline-git-hxwkm] {"level":"info","ts":1585623744.5506163,"caller":"git/git.go:130","msg":"Successfully initialized and updated submodules in path /workspace/workspace"}

[file-exists : condition-check-file-exists] + test -f /workspace/workspace/README2.md

failed to get logs for task file-exists : container step-condition-check-file-exists has failed  : [{"key":"StartedAt","value":"2020-03-31T03:02:25Z","resourceRef":{}}]
Unable to get Taskrun: taskruns.tekton.dev "condtional-pr-then-check-bqdx4" not found
```

in those, the log of
```
task file-exists has failed: "step-condition-check-file-exists" exited with code 1 (image: "docker-pullable://alpine@sha256:b276d875eeed9c7d3f1cfa7edb06b22ed22b14219a7d67c52c56612330348239"); for logs run: kubectl -n default logs condtional-pr-then-check-bqdx4-file-exists-d8zzr-pod-qwqz6 -c step-condition-check-file-exists
```
is output from ```hasTaskRunFailed``` method.

And the log of 
```
Unable to get Taskrun: taskruns.tekton.dev "condtional-pr-then-check-bqdx4" not found
```
is output from not filtering tasks by ```HasScheduled``` method.

So I think at least ```HasScheduled``` should be added to ```without -f flag``` for now,
But I have no confidence whether it is best... 🤔

danielhelfand commented:
Sorry for the long pause here on this. I think there are some design considerations around Conditions that should be addressed in pipelines as opposed to putting the logic into `tkn`. My thought is that Conditions should be treated as a regular TaskRun that is part of a PipelineRun. `tkn` should be able to gather logging information in a consistent manner and shouldn't need to add in extra checks for if Conditions are present.

As of now, when a Condition TaskRun is created as a result of a PipelineRun, the resulting TaskRun isn't considered a TaskRun that is part of the PipelineRun. Why this Condition TaskRun isn't considered a TaskRun of the PipelineRun is an odd design choice that I think should be addressed. If the Condition itself manifests as a TaskRun, then it should be treated as one as part of a PipelineRun.

By doing so, our standard logging approach for TaskRuns would work as expected with Conditions.

Let's keep this open for now as there are good design considerations and tests you have put together, but I think the changes to address this should come from pipelines.
